win/WIN_Slider.cpp

```diff
@@ -0,0 +1,127 @@
+#include "WIN_Slider.h"
+
+using namespace win;
+
+Slider::Slider(SDL_Renderer* renderer, gfx::Rectangle rect, const char* name, gfx::Colour fillColour, gfx::Colour outlineColour, int const initialPos, int const slideMin, int const slideMax)
+	: UIelement(rect, name)
+	, renderer_(renderer)
+	, slideLineMin_(slideMin)
+	, slideLineMax_(slideMax)
+	, lineRect_(rect.x+2, int(rect.y+rect.height/2)-2, rect.width-5, 3)
+	, lineColour_({0,0,0,255})
+	, markerPos_(initialPos)
+	, markerRect_(initialPos, rect.y, 5, rect.height)
+	, markerColour_({ 0,0,0,255 })
+	, holdMarker_(false)
+
+{
+	setForegroundColour(fillColour);
+	setBackgroundColour(outlineColour);
+}
+
+void Slider::updateLineMarker()
+{
+	lineRect_ = gfx::Rectangle(getRect().x + 2, (getRect().y + getRect().height / 2) - 2, getRect().width - 5, 3);
+	markerRect_ = gfx::Rectangle(markerPos_, getRect().y, 5, getRect().height);
+}
+
+void Slider::positionFromValue(int const val)
+{
+	markerPos_ = lineRect_.x +  int(double(val) * double(lineRect_.width) / (double(slideLineMax_) - double(slideLineMin_)));
+	markerRect_.x = markerPos_;
+}
+
+// Need to convert scaling from the position
+int Slider::valueFromPosition() const
+{
+	const auto relativeX = markerPos_ - lineRect_.x;
+	return int(double(relativeX) * (double(slideLineMax_) - double(slideLineMin_)) / double(lineRect_.width));
+}
+
+void Slider::moveMarker()
+{
+	int xMouse = 0;
+	int yMouse = 0;
+
+	SDL_GetMouseState(&xMouse, &yMouse);
+	if (xMouse > lineRect_.x + lineRect_.width - 2) {
+		xMouse = lineRect_.x + lineRect_.width;
+	}
+	if (xMouse < lineRect_.x + 2) {
+		xMouse = lineRect_.x;
+	}
+	markerRect_.x = xMouse;
+	markerPos_ = xMouse;
+}
+
+void Slider::update()
+{
+	updateLineMarker();
+}
+
+
+void Slider::updateAndRerender()
+{
+	updateLineMarker();
+	
+	draw();
+	SDL_RenderPresent(renderer_);
```

Comment:
Again, this forces redraw of the entire screen which might happen anyway! Should be no need to draw() as that should be called when renderering.
**SHOULD**

Reply:
changed as above